staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go

```diff
@@ -175,7 +175,15 @@ type VersionedOptions struct {
 	//
 	// Optional.
 	RemovedVersion *version.Version
-
+	// FeatureEnabled returns true if these options are enabled by feature gates,
+	// and returns false if these options are not enabled due to feature gates.
+	//
+	// This takes priority over IntroducedVersion / RemovedVersion for the NewExpressions environment.
+	//
+	// The StoredExpressions environment ignores this function.
+	//
+	// Optional.
+	FeatureEnabled func() bool
```

Comment:
@jpbetz, this was what I came up with to make feature-gate-informed VersionedOptions.

StoredExpressions assumes gated libraries are always available and *only* goes off of IntroducedVersion / RemovedVersion, since stuff in storage could have gotten in when a feature gate was enabled.

Making it override IntroducedVersion / RemovedVersion for NewExpressions allows you to opt into an alpha feature and use the CEL bits for it right away (similar to how opting into alpha features lets you do other things that aren't n-1 compatible right away, like validation ratcheting).

Reply:
> StoredExpressions assumes gated libraries are always available and only goes off of IntroducedVersion / RemovedVersion, since stuff in storage could have gotten in when a feature gate was enabled.

Perfect.

> Making it override IntroducedVersion / RemovedVersion for NewExpressions allows you to opt into an alpha feature and use the CEL bits for it right away (similar to how opting into alpha features lets you do other things that aren't n-1 compatible right away, like validation ratcheting).

I like how this solves both for feature gating and for immediate use of CEL libraries for alpha features (DRA has also needed this).

Reply:
Does that mean that I can remove the `IntroducedVersion: version.MajorMinor(1, 0),` hack once this lands and I do... something else instead? :sweat_smile:

Reply:
Yep, this will cover the DRA use case (using a new CEL library immediately when introducing alpha features) as well as the need to support feature gates.  @pohly can you leave a link to the DRA code so I can be sure to follow up?